### PR TITLE
Private Jet Mode feature

### PR DESCRIPTION
**Pull request by pubudu538**
### Proposed changes in this pull request

This PR address https://github.com/wso2/product-apim/issues/1831.

Unit tests are added.
DAO integration tests are added.
Fixes https://github.com/wso2/product-apim/issues/2780

### When should this PR be merged

ASAP. Need this for m19 release.

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIPublisherImpl.java [hunk 1]
```diff
@@ -261,6 +264,90 @@ public API getAPIbyUUID(String uuid) throws APIManagementException {
...
+        API api = getAPIbyUUID(apiId);
+        try {
+
+            validateChangingDedicatedGateway(api);
+            // label generation for the API
+            String autoGenLabelName = ContainerBasedGatewayConstants.PER_API_GATEWAY_PREFIX + apiId;
+            List<String> labelSet = new ArrayList<>();
+
+            if (dedicatedGateway.isEnabled()) {
+
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIPublisherImpl.java]: do we require the line spaces?

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/api/APIGateway.java [hunk 2]
```diff
@@ -18,6 +18,7 @@
 */
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/api/APIGateway.java]: We will require to improve the debug logs. Can we add more debug logs for troubleshooting purposes?

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIStoreImpl.java [hunk 3]
```diff
@@ -1824,4 +1825,63 @@ public void removeObserver(EventObserver observer) {
...
+                if (label == null) {
+                    // create a label
+                    List<Label> labelList = new ArrayList<>();
+                    List<String> accessUrls = new ArrayList<>();
+                    accessUrls.add(APIMgtConstants.HTTPS + APIMgtConstants.WEB_PROTOCOL_SUFFIX + autoGenLabelName);
+                    Label autoGenLabel = new Label.Builder().id(UUID.randomUUID().toString()).
+                            name(autoGenLabelName).accessUrls(accessUrls).type(APIMgtConstants.LABEL_TYPE_GATEWAY)
+                            .build();
+                    labelList.add(autoGenLabel);
+                    //Add to the db
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIStoreImpl.java]: i dont think we need this comment

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIStoreImpl.java [hunk 4]
```diff
@@ -1824,4 +1825,63 @@ public void removeObserver(EventObserver observer) {
...
+            // label generation for the API
+            String autoGenLabelName = ContainerBasedGatewayConstants.PER_API_GATEWAY_PREFIX + apiId;
+            List<String> labelSet = new ArrayList<>();
+
+            if (dedicatedGateway.isEnabled()) {
+
+                Label label = getLabelDAO().getLabelByName(autoGenLabelName);
+
+                if (label == null) {
+                    // create a label
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIStoreImpl.java]: may be more specific comment would be grate

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/KubernetesGatewayImpl.java [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+        HasMetadata resource = getResourcesFromTemplate(serviceTemplate);
+
+        try {
+            if (resource instanceof Service) {
+                // check whether there are existing service already
+                if (client.services().inNamespace(namespace).withName(serviceName).get() == null) {
+                    Service service = (Service) resource;
+                    Service result = client.services().inNamespace(namespace).create(service);
+                    if (log.isDebugEnabled()) {
+                        log.debug("CMS Type: " + cmsType + ". The Service Definition : " + serviceTemplate);
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/KubernetesGatewayImpl.java]: can we follow same logging format where we use {} and list of key words which will be the standradr of our logging

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/ContainerBasedGatewayTemplateBuilder.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+    /**
+     * Set velocity context for Gateway Service.
+     *
+     * @param templateValues VelocityContext template values Map
+     * @return Velocity Context Object
+     */
+    private VelocityContext setVelocityContextValues(Map<String, String> templateValues) {
+
+        VelocityContext context = new VelocityContext();
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/ContainerBasedGatewayTemplateBuilder.java]: incase of if we need to add more details for the generated velocity template, please check all information required is ther e in the veloity context.

Comment (pubudu538) [reply components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/template/ContainerBasedGatewayTemplateBuilder.java]: I have added the required values for the context. For Kubernetes specific values, we can directly change the templates

File: components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIPublisherImpl.java (hunk 1 above)
Comment (pubudu538) [reply components/apimgt/org.wso2.carbon.apimgt.core/src/main/java/org/wso2/carbon/apimgt/core/impl/APIPublisherImpl.java]: It is more cleaner.

**Pull request merged**